Add doc on resource reconciling

mrIncompetent opened the pull request:
**Release note**:
```release-note
NONE
```

mrIncompetent committed:
add doc on resource reconciling

mrIncompetent commented:
Once we migrated completely to the controller-runtime, this doc must be updated to reflect the removal of the informer.

mrIncompetent committed:
add TOC entry

alvaroaleman commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,123 @@
+
+## Reconciling
+
+Our reconciling is designed in a way, that every Object will be overwritten on each sync.
```
This already assumes familarity with some concepts, maybe something like `that we got notified whenever an object change, compare it to a desired state and if there is a diff update it`?

alvaroaleman commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,123 @@
+
+## EnsureNamedObject
+
+EnsureNamedObject is a generic "reconcile" function which will update the existing object or create it.
```
Plase use `EnsureNamedObject` to make sure its obvious this is code. Also I would link it to the code

mrIncompetent replied:
I'm not sure what you mean :thinking:

alvaroaleman replied:
replace

```
EnsureNamedObject is a generic "reconcile"
```

with

```
`EnsureNamedObject` is a generic "reconcile"
```

so it gets highlighted as code

alvaroaleman commented on docs/resource-handling.md:
````diff
@@ -0,0 +1,123 @@
+EnsureNamedObject is a generic "reconcile" function which will update the existing object or create it.
+If the existing object does not differ from the "wanted" object, EnsureNamedObject will issue any API call. 
+```go
+func EnsureNamedObject(name string, namespace string, rawcreate ObjectCreator, store informerStore, client ctrlruntimeclient.Client) error
````
I personally wouldn't nention `EnsureNamedObject` at all. It has a really weird signature, taking the store as argument. The only reason we need is is that the `cluster controller` uses some parts of `cluster-runtime` (dynamic cache, dynamic store), but doesn't use others (cache-backed dynamic client).

Can we instead just mention `EnsureNamedObjectV2` as that is what ppl should use?

mrIncompetent replied:
https://github.com/kubermatic/kubermatic/pull/2893#issuecomment-468622725

alvaroaleman replied:
> Once we migrated completely to the controller-runtime, this doc must be updated to reflect the removal of the informer.

I disagree. The ppl that need this doc will not work on the relevant parts in the cluster controller. And the cluster-controller is the only place where we are using `EnsureNamedObject`. Everything else should not use it.

alvaroaleman commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,123 @@
+
+## ObjectCreator
+
+ObjectCreator defines an interface to create/update a runtime.Object.
```
`is a function type...`

alvaroaleman reviewed (commented):
Some comments. Also what about just going through this with @nikhita and check how much she understand? We both are pretty deep into the topic so none of us has an unbiased view.

Also IMHO it would probably be helpful to do this more on a use-case basis, as in:

* If you want to add a new object, create a `rawCreator` for your type
* If you want to alter an existing resources, wrap the `rawCreator`

mrIncompetent committed:
pr feedback

nikhita commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,226 @@
+## EnsureNamedObject
+
+`EnsureNamedObject` is a generic "reconcile" function which will update the existing object or create it.
+If the existing object does not differ from the "wanted" object, `EnsureNamedObject` will issue any API call. 
```
nit: not issue any API call?

mrIncompetent replied:
Fixed with https://github.com/kubermatic/kubermatic/pull/2893/commits/cbaaf2882bb5a6445959916e02c4bb2a0faa8145

nikhita commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,226 @@
+		"password": []byte("bar"),
+	}
+
+	return existing, nil
```
For this one, can we do something like this to ensure proper handling of `nil`?

```go
secret := existing
if secret == nil {
   secret = &corev1.Secret{}
}

secret.Name = "wonderful-secret"
...
```

mrIncompetent replied:
I added a comment with https://github.com/kubermatic/kubermatic/pull/2893/commits/cbaaf2882bb5a6445959916e02c4bb2a0faa8145

nikhita commented on docs/resource-handling.md:
````diff
@@ -0,0 +1,226 @@
+}
+```
+
+### Reconcile your new resource
````
Can we just add a line that this reconciliation needs to go in the controller?

(because writing the creator func above isn't part of the controller and it just confused me for a second :see_no_evil: )

mrIncompetent replied:
Done with https://github.com/kubermatic/kubermatic/pull/2893/commits/cbaaf2882bb5a6445959916e02c4bb2a0faa8145

nikhita commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,226 @@
+
+For wrapping/modifying existing resources we have 2 options:
+- Pass `ObjectModifier` functions to the typed `Reconcile*` functions
+  Good if you wan't to modify all resources of a specific type
```
nit: s/wan't/want

mrIncompetent replied:
done with https://github.com/kubermatic/kubermatic/pull/2893/commits/cbaaf2882bb5a6445959916e02c4bb2a0faa8145

nikhita commented:
> Once we migrated completely to the controller-runtime, this doc must be updated to reflect the removal of the informer.

@mrIncompetent IIUC this doc is not meant to reflect reconciliation using controller-runtime...or is it? If yes, I think it would be useful to write down how to use the generated functions by controller-runtime (in brief).

nikhita commented:
Also.... can we also talk a bit about how funcs like `userClusterEnsureConfigMaps` funcs work? I think the gist is somewhat covered in the doc but could we also say something in more explicit terms (in much nicer words of course :)):

- it will use `data` for the creators and create an object (say `cm`).
- it will then use the client to `GET` an object (of the same type) with that name (say `cm.Name`)
- if it doesn't find such an object, it will create it using the client
- we then deepcopy the object to make sure we don't mutate it
- compare with the existing object created in the first point -> if it is equal, don't do anything. if not, update the object.

When I first saw this code, it took me a while to understand why we were using `Creators`, etc and @alvaroaleman walked me through these funcs which made the _why_ part much clearer :grimacing:

nikhita commented on docs/resource-handling.md:
```diff
@@ -0,0 +1,226 @@
+
+## Reconciling
+
+Our reconciling is designed in a way, that we got notified whenever an object change, compare it to a desired state and if there is a diff update it.
```
re desired state...can we say that we get the desired state using the creators?

mrIncompetent committed:
more review feedback

mrIncompetent commented:
> Also.... can we also talk a bit about how funcs like userClusterEnsureConfigMaps funcs work? I think the gist is somewhat covered in the doc but could we also say something in more explicit terms (in much nicer words of course :)):

`userClusterEnsureConfigMaps` should be replaced by `ReconcileConfigMaps`.
There shall be no custom reconciling implementation. 
We only introduced `userClusterEnsureConfigMaps` because we had no way of writing a generic reconciling function which covers plain HTTP + Cache usage - As all code which reconciled user cluster resources had no cache.

mrIncompetent commented:
> @mrIncompetent IIUC this doc is not meant to reflect reconciliation using controller-runtime...or is it? If yes, I think it would be useful to write down how to use the generated functions by controller-runtime (in brief).

The above mentioned functions can be used by controllers using the controller-runtime.
Only caveat is that each reconciler needs to have the "informerFactory":
```go
informerFactory := mgr.GetCache()
```
As its a required parameter for the `Reconcile*` functions - but that should drop as soon as i migrated the cluster+monitoring controller (which i started to do)

mrIncompetent committed:
add TOC, info about migrations & better example code

mrIncompetent committed:
add info on change of Renconcile* functions

mrIncompetent commented:
@nikhita @alvaroaleman 
I updated the doc to include:
- A TOC
- Example code which covers the integration in legacy controllers + controller-runtime controllers
- A info on a potential change of the function signature of `EnsureNamedObject` + the typed `Reconcile*` functions.
- Added a info which tells to prefer the typed `Reconcile*` functions over the `EnsureNamedObject`

mrIncompetent commented:
Merging this one as there has been no comment in the last 8 days and having it in is more useful than an open PR...
/approve
/lgtm

nikhita commented:
/retest

nikhita commented:
/retest

nikhita commented:
/retest

mrIncompetent commented:
/retest

mrIncompetent commented:
/retest

nikhita commented:
/retest

:/

mrIncompetent commented:
/retest

mrIncompetent commented:
oh come on
/retest

nikhita commented:
Worst case, if it flakes again, _mayyybeee_ you can do a `/override <test-name>` since this PR just has docs changes?

But I don't want this to set a precedent for using that command too. :see_no_evil:

mrIncompetent commented:
@nikhita oh you shouldn't have told me that :smiling_imp:
